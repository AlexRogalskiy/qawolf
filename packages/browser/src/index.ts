export { Page as PuppeteerPage } from "puppeteer";
export * from "./browser";
export * from "./find";
export * from "./retry";
export * from "./strokes";
