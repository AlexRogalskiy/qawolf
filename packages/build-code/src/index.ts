export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
