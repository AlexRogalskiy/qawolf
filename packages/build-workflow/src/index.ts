// export * from "./buildClickSteps";
// export * from "./buildScrollSteps";
// export * from "./buildSelectSteps";
// export * from "./buildSteps";
// export * from "./buildTypeSteps";
// export * from "./buildWorkflow";
