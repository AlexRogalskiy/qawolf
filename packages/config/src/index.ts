import dotenv from "dotenv";

const dotEnvPath = process.env.DOTENV_PATH;
dotenv.config(dotEnvPath ? { path: dotEnvPath } : {});

const testUrl = process.env.TEST_URL || "http://localhost:5000/";

const parseBool = (value: string | undefined) => {
  return !!value && value !== "false";
};

const parseNumber = (value: string | undefined) => {
  if (typeof value === "undefined") return value;

  return parseFloat(value);
};

export const CONFIG = {
  chromeExecutablePath: process.env.CHROME_EXECUTABLE_PATH,
  dataAttribute: process.env.DATA_ATTRIBUTE || null,
  headless: parseBool(process.env.HEADLESS),
  keepBrowserOpen: parseBool(process.env.KEEP_BROWSER_OPEN),
  logPath: process.env.LOG_PATH,
  saveBrowserWsPath: process.env.SAVE_BROWSER_WS_PATH,
  screenshotPath: process.env.SCREENSHOT_PATH,
  sleepAfterEach: parseNumber(process.env.SLEEP_AFTER_EACH),
  testUrl,
  useLocalModule: parseBool(process.env.USE_LOCAL_MODULE)
};
