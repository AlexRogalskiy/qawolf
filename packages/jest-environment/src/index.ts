import { RunnerEnvironment } from "./RunnerEnvironment";
import "./types";

module.exports = RunnerEnvironment;
