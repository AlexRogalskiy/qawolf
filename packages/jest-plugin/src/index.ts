// setup fail-fast since we do not want to continue when a step fails
// hopefully jest-circus implements this behavior, but was not the case when I tested jest 24.9.0
// https://github.com/facebook/jest/issues/2867#issuecomment-476647864
// https://github.com/facebook/jest/issues/6527#issuecomment-506278251

type Callback = (name: string) => void;

const j = jasmine as any;

// require since there are no types
const failFast = require("@qawolf/jasmine-fail-fast");
const jasmineEnv = j.getEnv();
jasmineEnv.addReporter(failFast.init());

// use Jasmine specStarted reporter until we can use Jest Circus
// https://github.com/qawolf/qawolf/issues/345
j.qaw_callbacks = [];

j.qaw_onSpecStarted = (callback: Callback) => j.qaw_callbacks.push(callback);

jasmineEnv.addReporter({
  specStarted: (result: any) => {
    j.qaw_callbacks.forEach((cb: Callback) => cb(result.description));
  }
});
