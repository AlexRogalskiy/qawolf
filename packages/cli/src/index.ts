#!/usr/bin/env node

import { buildTest } from "@qawolf/build-test";
// import { buildWorkflow } from "@qawolf/build-workflow";
import { Browser } from "@qawolf/browser";
import { CONFIG } from "@qawolf/config";
import { logger } from "@qawolf/logger";
import { Runner } from "@qawolf/runner";
import { Workflow } from "@qawolf/types";
import program from "commander";
import { outputFile, outputJson, readJson } from "fs-extra";
import { snakeCase } from "lodash";
import { resolve } from "path";
import { runTest } from "./runTest";
import { parseUrl, getUrlRoot } from "./utils";

// program
//   .command("build <eventsPath> <name>")
//   .description("build a test from events")
//   .action(async (eventsPath, name) => {
//     const sourcePath = resolve(eventsPath);
//     logger.verbose(`read events from ${sourcePath}`);
//     const events = await readJson(sourcePath);

//     const destPath = `${process.cwd()}/.qawolf`;
//     const formattedName = snakeCase(name);
//     const destWorkflowPath = `${destPath}/workflows/${formattedName}.json`;
//     const destTestPath = `${destPath}/tests/${formattedName}.test.js`;

//     logger.verbose(`build workflow -> ${destTestPath}`);
//     const workflow = buildWorkflow(events, formattedName);
//     await outputJson(destWorkflowPath, workflow, { spaces: " " });

//     logger.verbose(`build test -> ${destTestPath}`);
//     const test = buildTest(workflow);
//     await outputFile(destTestPath, test, "utf8");

//     process.exit(0);
//   });

let recordCommand = program
  .command("record <url> [name]")
  .description("record a workflow and create a test");

if (CONFIG.development) {
  recordCommand = recordCommand.option(
    "-e, --events",
    "record events (for debugging)"
  );
}

recordCommand.action(async (urlArgument, optionalName, cmd) => {
  const url = parseUrl(urlArgument);
  const name = optionalName || getUrlRoot(url);
  logger.verbose(`record url "${url.href}"`);

  const browser = await Browser.create({ record: true, url: url.href });
  await browser.waitForClose();

  const destPath = `${process.cwd()}/.qawolf`;
  const formattedName = snakeCase(name);

  if (cmd.events) {
    const destEventsPath = `${destPath}/events/${formattedName}.json`;
    logger.verbose(`create events "${name}" -> ${destEventsPath}`);
    await outputJson(destEventsPath, browser.events, { spaces: " " });
  } else {
    logger.verbose(`create test "${name}"`);
  }

  process.exit(0);
});

program
  .command("run [name]")
  .description("run a workflow")
  .action(async name => {
    const workflowPath = `${process.cwd()}/.qawolf/workflows/${snakeCase(
      name
    )}.json`;
    const workflow = (await readJson(workflowPath)) as Workflow;
    const runner = await Runner.create(workflow);

    await runner.run();
    await runner.close();

    process.exit(0);
  });

program
  .command("test [name]")
  .description("run a test")
  .action(async name => {
    const results = await runTest(name ? snakeCase(name) : null);
    const success = results.numFailedTestSuites < 1;

    process.exit(success ? 0 : 1);
  });

program.allowUnknownOption(false);

program.parse(process.argv);

if (!process.argv.slice(2).length) {
  console.log("\n");
  program.outputHelp();
}
