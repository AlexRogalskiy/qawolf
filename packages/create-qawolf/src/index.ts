#!/usr/bin/env node
import { install as installCi } from 'playwright-ci';
import {
  logError,
  logInstallDependencies,
  logUseTypeScript,
  promptRootDir,
} from './cli';
import { detectTypeScript, writeConfig } from './config';
import {
  addDevDependencies,
  readPackageJson,
  installDependencies,
} from './packageJson';

(async (): Promise<void> => {
  try {
    const useYarn = process.argv[process.argv.length - 1] === '--yarn';

    // create a new line for yarn create
    console.log();

    // run this first to ensure package.json
    await readPackageJson();

    const rootDir = await promptRootDir();

    const useTypeScript = await detectTypeScript();
    logUseTypeScript(useTypeScript);

    await installCi(true);

    await writeConfig({ rootDir, useTypeScript });

    const packages = await addDevDependencies(useTypeScript);

    logInstallDependencies(packages, useYarn);

    installDependencies(useYarn);
  } catch (error) {
    logError(error);
    process.exit(1);
  }
})();
