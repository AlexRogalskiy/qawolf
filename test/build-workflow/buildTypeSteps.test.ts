import { loadFixtures } from '../loadFixtures';
import { buildTypeSteps } from '../../src/build-workflow';

describe('buildTypeSteps', () => {
  it('builds expected steps for login', async () => {
    const events = (await loadFixtures('login')).events;
    const steps = buildTypeSteps(events);
    expect(steps).toMatchSnapshot();
  });

  it('builds expected steps for selectall', async () => {
    const events = (await loadFixtures('textInputsSelectAll')).events;
    const steps = buildTypeSteps(events);
    expect(steps).toMatchSnapshot();
  });

  // TODO
  // it('builds expected steps for hotkeys', async () => {
  //   const events = await loadEvents('githubIssuesShortcut');
  //   const steps = buildTypeSteps(events);
  //   expect(steps).toMatchSnapshot();
  // });
});
