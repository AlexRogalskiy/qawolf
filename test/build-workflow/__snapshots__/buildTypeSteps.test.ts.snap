// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTypeSteps builds expected steps for login 1`] = `
Array [
  Object {
    "action": "type",
    "cssSelector": undefined,
    "htmlSelector": "<div id=\\"root\\" qaw_innertext=\\"To log in, use tomsmith as the username, and SuperSecretPassword! as the password. Username Password Log in\\"><form qaw_innertext=\\"Username Password Log in\\"><input autocomplete=\\"off\\" id=\\"username\\" type=\\"text\\" value=\\"\\" qaw_labels=\\"Username\\" /></form></div>",
    "index": 4,
    "page": 0,
    "target": Object {
      "attrs": Object {
        "autocomplete": "off",
        "id": "username",
        "qaw_labels": "Username",
        "type": "text",
        "value": "",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "value": "tomsmith",
  },
  Object {
    "action": "type",
    "cssSelector": undefined,
    "htmlSelector": "<div id=\\"root\\" qaw_innertext=\\"To log in, use tomsmith as the username, and SuperSecretPassword! as the password. Username Password Log in\\"><form qaw_innertext=\\"Username Password Log in\\"><input autocomplete=\\"off\\" id=\\"username\\" type=\\"text\\" value=\\"tomsmith\\" qaw_labels=\\"Username\\" /></form></div>",
    "index": 20,
    "page": 0,
    "target": Object {
      "attrs": Object {
        "autocomplete": "off",
        "id": "username",
        "qaw_labels": "Username",
        "type": "text",
        "value": "tomsmith",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "value": "↓Tab↑Tab",
  },
  Object {
    "action": "type",
    "cssSelector": undefined,
    "htmlSelector": "<div id=\\"root\\" qaw_innertext=\\"To log in, use tomsmith as the username, and SuperSecretPassword! as the password. Username Password Log in\\"><form qaw_innertext=\\"Username Password Log in\\"><input autocomplete=\\"off\\" id=\\"password\\" type=\\"password\\" value=\\"\\" qaw_labels=\\"Password\\" /></form></div>",
    "index": 24,
    "page": 0,
    "target": Object {
      "attrs": Object {
        "autocomplete": "off",
        "id": "password",
        "qaw_labels": "Password",
        "type": "password",
        "value": "",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "value": "SuperSecretPassword!",
  },
  Object {
    "action": "type",
    "cssSelector": undefined,
    "htmlSelector": "<div id=\\"root\\" qaw_innertext=\\"To log in, use tomsmith as the username, and SuperSecretPassword! as the password. Username Password Log in\\"><form qaw_innertext=\\"Username Password Log in\\"><input autocomplete=\\"off\\" id=\\"password\\" type=\\"password\\" value=\\"SuperSecretPassword!\\" qaw_labels=\\"Password\\" /></form></div>",
    "index": 27,
    "page": 0,
    "target": Object {
      "attrs": Object {
        "autocomplete": "off",
        "id": "password",
        "qaw_labels": "Password",
        "type": "password",
        "value": "SuperSecretPassword!",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "value": "↓Enter↑Enter",
  },
]
`;

exports[`buildTypeSteps builds expected steps for selectall 1`] = `
Array [
  Object {
    "action": "type",
    "cssSelector": "[data-qa='html-text-input-filled']",
    "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input-filled\\" placeholder=\\"Filled text input\\" type=\\"text\\" value=\\"initial text\\" /></div></div>",
    "index": 6,
    "page": 0,
    "replace": true,
    "target": Object {
      "attrs": Object {
        "data-qa": "html-text-input-filled",
        "placeholder": "Filled text input",
        "type": "text",
        "value": "initial text",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "value": "hotkeys",
  },
  Object {
    "action": "type",
    "cssSelector": "[data-qa='html-text-input-filled']",
    "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input-filled\\" placeholder=\\"Filled text input\\" type=\\"text\\" value=\\"initial text\\" /></div></div>",
    "index": 25,
    "page": 0,
    "replace": true,
    "target": Object {
      "attrs": Object {
        "data-qa": "html-text-input-filled",
        "placeholder": "Filled text input",
        "type": "text",
        "value": "initial text",
      },
      "children": Array [],
      "name": "input",
      "type": "tag",
      "voidElement": true,
    },
    "value": "clickselectall",
  },
]
`;
